File: BlueJ-Projekt/Bedarfsampel.java
/**
 * Bedarfsampel auf dem Calliope mini V3.
 *
 * Hardware:
 *   - Straßenampel: linke RGB-LED (rot/gelb/grün)
 *   - Fußgängerampel: rechte RGB-LED (rot/grün)
 *   - Anforderungstaster: Button A
 */
public class Bedarfsampel {
    private final int S_LED = 0; // RGB-LED 0 = Straßenampel
    private final int F_LED = 2; // RGB-LED 2 = Fußgängerampel
    private int loopID;
    private CalliopeIO calliope;

    public Bedarfsampel() {
        calliope = new CalliopeIO();
        loopID = -1;
    }

    public void start() {
        if (loopID == -1) {
            calliope.open();
            calliope.clearPressed();
            sGruen();   // Straße: grün
            fRot();     // Fußgänger: rot
            loopID = calliope.startLoop(this, "loop");
        } else {
            System.out.println("Loop bereits gestartet");
        }
    }

    private void loop() {
        if (calliope.wasButtonAPressed()) {
            schaltePhase();
        }
    }

    public void stopp() {
        if (loopID != -1) {
            calliope.stopLoop(loopID);
            loopID = -1;
            calliope.clearRgb();
            calliope.close();
        }
    }

    private void schaltePhase() {
        calliope.pause(500);
        sGelb();                                // Straße: gelb
        calliope.pause(1000);
        sRot();                                 // Straße: rot
        calliope.pause(500);
        fGruen();                               // Fußgänger: grün
        calliope.pause(4000);
        fRot();                                 // Fußgänger: rot
        calliope.clearPressed();
        calliope.pause(500);
        sGelb();                                // Straße: rot+gelb (nur gelb in RGB-Darstellung)
        calliope.pause(500);
        sGruen();                               // Straße: grün
        calliope.pause(2000);
    }

    private void sRot()   { calliope.setRgb(S_LED, 255,   0,   0); }
    private void sGelb()  { calliope.setRgb(S_LED, 255, 200,   0); }
    private void sGruen() { calliope.setRgb(S_LED,   0, 255,   0); }
    private void fRot()   { calliope.setRgb(F_LED, 255,   0,   0); }
    private void fGruen() { calliope.setRgb(F_LED,   0, 255,   0); }
}

File: Software/Python/Beispiele/bedarfsampel.py
"""Bedarfsampel auf dem Calliope mini V3.

Hardware:
  - Straßenampel:    linke RGB-LED (rot/gelb/grün)
  - Fußgängerampel:  rechte RGB-LED (rot/grün)
  - Anforderungstaster: Button A
"""

import sys
sys.path.insert(0, "../Source")

from calliope_io import CalliopeIO

STRASSE = 0   # linke RGB-LED
FUSS = 2      # rechte RGB-LED

ROT = (255, 0, 0)
GELB = (255, 200, 0)
GRUEN = (0, 255, 0)


class Bedarfsampel:
    def __init__(self) -> None:
        self.c = CalliopeIO()
        self._loop_id = -1

    def start(self) -> None:
        if self._loop_id != -1:
            print("Loop bereits gestartet")
            return
        self.c.open()
        self.c.clear_pressed()
        self.c.set_rgb(STRASSE, *GRUEN)
        self.c.set_rgb(FUSS, *ROT)
        self._loop_id = self.c.start_loop(self._tick, delay_ms=50)

    def stopp(self) -> None:
        if self._loop_id == -1:
            return
        self.c.stop_loop(self._loop_id)
        self._loop_id = -1
        self.c.close()

    def _tick(self) -> None:
        if self.c.was_button_a_pressed():
            self._phase_durchlaufen()

    def _phase_durchlaufen(self) -> None:
        self.c.pause(500)
        self.c.set_rgb(STRASSE, *GELB)
        self.c.pause(1000)
        self.c.set_rgb(STRASSE, *ROT)
        self.c.pause(500)
        self.c.set_rgb(FUSS, *GRUEN)
        self.c.pause(4000)
        self.c.set_rgb(FUSS, *ROT)
        self.c.clear_pressed()
        self.c.pause(500)
        self.c.set_rgb(STRASSE, *GELB)
        self.c.pause(500)
        self.c.set_rgb(STRASSE, *GRUEN)
        self.c.pause(2000)


if __name__ == "__main__":
    ampel = Bedarfsampel()
    ampel.start()
    try:
        input("Enter drücken zum Beenden...\n")
    finally:
        ampel.stopp()
